File: frontend/src/api/retreat.ts
import { get, post, put, del, patch } from '@/api/client';
import {
  RetreatAccommodationResponse,
  RetreatHeadcountResponse,
  SuspendedMealListResponse,
  SuspendedMealReviewRequest,
  SuspendedMealStats,
  RetreatActiveResponse,
  RetreatCreateBody,
  RetreatCreateResponse,
  RetreatUpdateBody,
  BusCreateBody,
  ResearchListResponse,
  VehicleMemberListResponse,
  VehicleDashboardData,
  PatientRoomListResponse,
  PatientRoomReviewRequest,
  PatientRoomStats,
} from '@/models/retreat.types';

// ── 서스펜디드밀 ──────────────────────────────────────────────────────────────

export async function fetchSuspendedMealList(params?: {
  page?: number;
  size?: number;
  review_status?: string;
}): Promise<SuspendedMealListResponse> {
  return get<SuspendedMealListResponse>('/api/retreat/suspended-meal', params);
}

export async function fetchSuspendedMealStats(): Promise<SuspendedMealStats> {
  return get<SuspendedMealStats>('/api/retreat/suspended-meal/stats');
}

export async function reviewSuspendedMeal(
  applicationId: number,
  body: SuspendedMealReviewRequest
): Promise<void> {
  return put<void>(`/api/retreat/suspended-meal/${applicationId}/review`, body);
}

// ── 수련회 설정 ───────────────────────────────────────────────────────────────

export async function getActiveRetreat(): Promise<RetreatActiveResponse> {
  return get<RetreatActiveResponse>('/api/retreat/active');
}

export async function createRetreat(body: RetreatCreateBody): Promise<RetreatCreateResponse> {
  return post<RetreatCreateResponse>('/api/retreat', body);
}

export async function updateRetreat(retreatId: number, body: RetreatUpdateBody): Promise<void> {
  return put<void>(`/api/retreat/${retreatId}`, body);
}

export async function createBus(body: BusCreateBody): Promise<{ bus_id: number }> {
  return post<{ bus_id: number }>('/api/bus', body);
}

export async function deleteBus(busId: number): Promise<void> {
  return del<void>(`/api/bus/${busId}`);
}

export async function completeRetreat(retreatId: number): Promise<void> {
  return put<void>(`/api/retreat/${retreatId}/complete`);
}

// ── 수련회 인원조사 ───────────────────────────────────────────────────────────

export async function fetchRetreatHeadcount(): Promise<RetreatHeadcountResponse> {
  return get<RetreatHeadcountResponse>('/api/retreat/headcount');
}

export async function fetchResearchList(): Promise<ResearchListResponse> {
  return get<ResearchListResponse>('/api/retreat/research/list');
}

export async function patchResearchFeePaid(memberId: number, isFeePaid: boolean): Promise<void> {
  return patch<void>(`/api/retreat/research/${memberId}/fee-paid`, { is_fee_paid: isFeePaid });
}

export async function fetchVehicleMemberList(params?: { gyogu?: number; team?: number; bus_id?: number }): Promise<VehicleMemberListResponse> {
  const query = new URLSearchParams();
  if (params?.gyogu)  query.set('gyogu',  String(params.gyogu));
  if (params?.team)   query.set('team',   String(params.team));
  if (params?.bus_id) query.set('bus_id', String(params.bus_id));
  const qs = query.toString();
  return get<VehicleMemberListResponse>(`/api/retreat/vehicle-members${qs ? `?${qs}` : ''}`);
}

// ── 환자방 ────────────────────────────────────────────────────────────────────

export async function fetchPatientRoomList(params?: {
  page?: number;
  size?: number;
  review_status?: string;
}): Promise<PatientRoomListResponse> {
  return get<PatientRoomListResponse>('/api/retreat/patient-room', params);
}

export async function fetchPatientRoomStats(): Promise<PatientRoomStats> {
  return get<PatientRoomStats>('/api/retreat/patient-room/stats');
}

export async function reviewPatientRoom(applicationId: number, body: PatientRoomReviewRequest): Promise<void> {
  return put<void>(`/api/retreat/patient-room/${applicationId}/review`, body);
}

// ── 수련회 차량조사 ───────────────────────────────────────────────────────────

export async function fetchRetreatVehicle(): Promise<VehicleDashboardData> {
  return get<VehicleDashboardData>('/api/retreat/vehicle');
}

// ── 수련회 숙소인원 ───────────────────────────────────────────────────────────

export async function fetchRetreatAccommodation(params?: {
  gyogu_no?: number;
  team_no?: number;
}): Promise<RetreatAccommodationResponse> {
  return get<RetreatAccommodationResponse>('/api/retreat/accommodation', params);
}

File: backend/app/api/v1/retreat/retreat.py
"""수련회 설정 API — 생성 / 조회 / 수정"""
from typing import Optional

from fastapi import APIRouter, Depends, Path, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.retreat import (
    RetreatCreate, RetreatUpdate, RetreatCreateResponse, RetreatActiveResponse,
    ResearchListResponse, VehicleMemberListResponse, VehicleDashboardResponse,
    RetreatHeadcountResponse, RetreatAccommodationResponse,
    AdminSuspendedMealListResponse, AdminSuspendedMealStats, AdminSuspendedMealReviewRequest,
)
from app.services.retreat import (
    svc_create_retreat, svc_get_active_retreat, svc_update_retreat, svc_complete_retreat,
    svc_get_research_list, svc_get_vehicle_member_list, svc_get_vehicle_dashboard,
    svc_get_headcount, svc_get_accommodation,
    svc_get_admin_suspended_meal_list, svc_get_admin_suspended_meal_stats, svc_review_suspended_meal,
)

router = APIRouter()


@router.post("", response_model=RetreatCreateResponse, status_code=201, summary="수련회 생성")
def create_retreat(body: RetreatCreate, db: Session = Depends(get_db)):
    return svc_create_retreat(db, body)


@router.get("/active", response_model=RetreatActiveResponse, summary="현재 수련회 + 버스 목록 조회")
def get_active_retreat(db: Session = Depends(get_db)):
    return svc_get_active_retreat(db)


@router.get("/headcount", response_model=RetreatHeadcountResponse, summary="인원조사 집계")
def get_headcount(db: Session = Depends(get_db)):
    return svc_get_headcount(db)


@router.get("/accommodation", response_model=RetreatAccommodationResponse, summary="숙소/야식 인원 집계")
def get_accommodation(
    gyogu_no: Optional[int]  = Query(None),
    team_no:  Optional[int]  = Query(None),
    is_imwondan: Optional[bool] = Query(None),
    db: Session = Depends(get_db),
):
    return svc_get_accommodation(db, gyogu_no, team_no)


@router.get("/vehicle", response_model=VehicleDashboardResponse, summary="대시보드 차량탭 집계")
def get_vehicle_dashboard(db: Session = Depends(get_db)):
    return svc_get_vehicle_dashboard(db)


@router.get("/research/list", response_model=ResearchListResponse, summary="인원조사 명단 (관리자)")
def get_research_list(
    gyogu: Optional[int] = Query(None),
    team: Optional[int] = Query(None),
    survey_status: Optional[str] = Query(None),
    db: Session = Depends(get_db),
):
    return svc_get_research_list(db, gyogu, team, survey_status)


@router.get("/vehicle-members", response_model=VehicleMemberListResponse, summary="차량조사 명단 (관리자)")
def get_vehicle_member_list(
    gyogu: Optional[int] = Query(None),
    team: Optional[int] = Query(None),
    db: Session = Depends(get_db),
):
    return svc_get_vehicle_member_list(db, gyogu, team)


@router.get("/suspended-meal", response_model=AdminSuspendedMealListResponse, summary="서스펜디드밀 신청 목록 (관리자)")
def get_admin_suspended_meal_list(
    page:          int           = Query(1, ge=1),
    size:          int           = Query(20, ge=1, le=100),
    review_status: Optional[str] = Query(None),
    db: Session = Depends(get_db),
):
    return svc_get_admin_suspended_meal_list(db, review_status, page, size)


@router.get("/suspended-meal/stats", response_model=AdminSuspendedMealStats, summary="서스펜디드밀 통계 (관리자)")
def get_admin_suspended_meal_stats(db: Session = Depends(get_db)):
    return svc_get_admin_suspended_meal_stats(db)


@router.put("/suspended-meal/{application_id}/review", status_code=200, summary="서스펜디드밀 승인/반려")
def review_suspended_meal(
    application_id: int = Path(...),
    body: AdminSuspendedMealReviewRequest = ...,
    db: Session = Depends(get_db),
):
    svc_review_suspended_meal(db, application_id, body)
    return {"ok": True}


@router.put("/{retreat_id}", status_code=200, summary="수련회 기본 정보 수정")
def update_retreat(
    retreat_id: int = Path(...),
    body: RetreatUpdate = ...,
    db: Session = Depends(get_db),
):
    svc_update_retreat(db, retreat_id, body)
    return {"ok": True}


@router.put("/{retreat_id}/complete", status_code=200, summary="수련회 완료 처리")
def complete_retreat(retreat_id: int = Path(...), db: Session = Depends(get_db)):
    svc_complete_retreat(db, retreat_id)
    return {"ok": True}
